consul/server_manager/server_manager.go

```diff
@@ -0,0 +1,340 @@
+	serverCfg := sm.getServerConfig()
+
+	// Remove the server if known
+	n := len(serverCfg.servers)
+	for i := 0; i < n; i++ {
+		if serverCfg.servers[i].Name == server.Name {
+			newServers := make([]*server_details.ServerDetails, 0, len(serverCfg.servers)-1)
+			newServers = append(newServers, serverCfg.servers[:i]...)
+			newServers = append(newServers, serverCfg.servers[i+1:]...)
+			serverCfg.servers = newServers
+
+			sm.saveServerConfig(serverCfg)
+			return
+		}
+	}
+}
+
+// refreshServerRebalanceTimer is only called once the rebalanceTimer
+// expires.  Historically this was an expensive routine and is intended to be
+// run in isolation in a dedicated, non-concurrent task.
+func (sm *ServerManager) refreshServerRebalanceTimer(timer *time.Timer) time.Duration {
+	serverCfg := sm.getServerConfig()
+	numConsulServers := len(serverCfg.servers)
+	// Limit this connection's life based on the size (and health) of the
+	// cluster.  Never rebalance a connection more frequently than
+	// connReuseLowWatermarkDuration, and make sure we never exceed
+	// clusterWideRebalanceConnsPerSec operations/s across numLANMembers.
+	clusterWideRebalanceConnsPerSec := float64(numConsulServers * newRebalanceConnsPerSecPerServer)
+	connReuseLowWatermarkDuration := clientRPCMinReuseDuration + lib.RandomStagger(clientRPCMinReuseDuration/clientRPCJitterFraction)
+	numLANMembers := sm.clusterInfo.NumNodes()
+	connRebalanceTimeout := lib.RateScaledInterval(clusterWideRebalanceConnsPerSec, connReuseLowWatermarkDuration, numLANMembers)
+
+	timer.Reset(connRebalanceTimeout)
+	return connRebalanceTimeout
+}
+
+// saveServerConfig is a convenience method which hides the locking semantics
+// of atomic.Value from the caller.
+func (sm *ServerManager) saveServerConfig(sc serverConfig) {
+	sm.serverConfigValue.Store(sc)
+}
+
+// Start is used to start and manage the task of automatically shuffling and
+// rebalancing the list of consul servers.  This maintenance only happens
+// periodically based on the expiration of the timer.  Failed servers are
+// automatically cycled to the end of the list.  New servers are appended to
+// the list.  The order of the server list must be shuffled periodically to
+// distribute load across all known and available consul servers.
+func (sm *ServerManager) Start() {
+	var rebalanceTimer *time.Timer = time.NewTimer(time.Duration(initialRebalanceTimeoutHours * time.Hour))
+	var rebalanceTaskDispatched int32
+
+	func() {
+		sm.serverConfigLock.Lock()
+		defer sm.serverConfigLock.Unlock()
+
+		serverCfgPtr := sm.serverConfigValue.Load()
+		if serverCfgPtr == nil {
+			panic("server config has not been initialized")
+		}
+		var serverCfg serverConfig
+		serverCfg = serverCfgPtr.(serverConfig)
+		sm.saveServerConfig(serverCfg)
+	}()
+
+	for {
+		select {
+		case <-rebalanceTimer.C:
+			sm.logger.Printf("[INFO] server manager: server rebalance timeout")
+			sm.RebalanceServers()
+
+			// Only run one rebalance task at a time, but do
```

Comment:
This was an artifact of the channel notification method where there could be many simultaneous server add/remove events happening that would cause Serf to backup.  This is no longer a possibility.